Sources/PackageModel/PackageIdentity.swift

```diff
@@ -0,0 +1,330 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import Foundation
+
+import TSCBasic
+import TSCUtility
+
+//@warn_unqualified_access
+/// When set to `false`,
+/// `PackageIdentity` uses the canonical location of package dependencies as its identity.
+/// Otherwise, only the last path component is used to identify package dependencies.
+public var _useLegacyIdentities: Bool = true {
+    willSet {
+        PackageIdentity.provider = newValue ? LegacyPackageIdentity.self : CanonicalPackageIdentity.self
+    }
+}
+
+internal protocol PackageIdentityProvider: LosslessStringConvertible {
```

Comment:
I think the identity provider also has to provide equality for package identities, not just an initializer. For the "real" new implementation of identity we will need a staggered comparison that's not just comparing strings.

Reply:
If we require `PackageIdentityProvider` to be `Equatable`, then it can only be used as a generic constraint, because it has `Self` requirements in `==`.

I found a way to work around this, but it's a good bit more complex than if we forego equatability:

```swift
public var _useLegacyIdentities: Bool = true {
    willSet {
        PackageIdentity.provider = newValue ? LegacyPackageIdentity.init : CanonicalPackageIdentity.init
    }
}

private typealias PackageIdentityProvider = (String) -> PackageIdentityProtocol

private protocol PackageIdentityProtocol: LosslessStringConvertible {
    init(_ string: String)
}

/// The canonical identifier for a package, based on its source location.
public struct PackageIdentity: LosslessStringConvertible, Hashable {
    fileprivate static var provider: PackageIdentityProvider = LegacyPackageIdentity.init
    public let description: String
    public init(_ string: String) {
        self.description = Self.provider(string).description
    }
}
```

Any strong preference between the two?

Reply:
I would just add a comparison function that's unrelated to `Equatable`. I think what we want to be able to do is compare two `PackageIdentity` instances using a comparison operation vended by `PackageIdentityProvider`.